src/receivers/HTTPModuleFunctions.ts

```diff
@@ -0,0 +1,264 @@
+      retryReason = retryReasonHeaderValue;
+    } else if (Array.isArray(retryReasonHeaderValue) && retryReasonHeaderValue.length > 0) {
+      // eslint-disable-next-line prefer-destructuring
+      retryReason = retryReasonHeaderValue[0];
+    }
+    return retryReason;
+  }
+
+  // ------------------------------------------
+  // HTTP request parsing and verification
+  // ------------------------------------------
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  public static parseHTTPRequestBody(req: BufferedIncomingMessage): any {
+    const bodyAsString = req.rawBody.toString();
+    const contentType = req.headers['content-type'];
+    if (contentType === 'application/x-www-form-urlencoded') {
+      const parsedQs = qsParse(bodyAsString);
+      const { payload } = parsedQs;
+      if (typeof payload === 'string') {
+        return JSON.parse(payload);
+      }
+      return parsedQs;
+    }
+    return JSON.parse(bodyAsString);
+  }
+
+  public static async parseAndVerifyHTTPRequest(
+    options: RequestVerificationOptions,
+    req: IncomingMessage,
+    _res?: ServerResponse,
+  ): Promise<BufferedIncomingMessage> {
+    const { signingSecret } = options;
+
+    // Consume the readable stream (or use the previously consumed readable stream)
+    const bufferedReq = await HTTPModuleFunctions.bufferIncomingMessage(req);
+
+    if (options.enabled !== undefined && !options.enabled) {
+      // As the validation is disabled, immediately return the bufferred reuest
+      return bufferedReq;
+    }
+
+    // Find the relevant request headers
+    const signature = HTTPModuleFunctions.getHeader(req, 'x-slack-signature');
+    const requestTimestampSec = Number(HTTPModuleFunctions.getHeader(req, 'x-slack-request-timestamp'));
+    if (Number.isNaN(requestTimestampSec)) {
+      throw new Error(
+        `${verifyErrorPrefix}: header x-slack-request-timestamp did not have the expected type (${requestTimestampSec})`,
+      );
+    }
+
+    // Calculate time-dependent values
+    const nowMsFn = options.nowMilliseconds ?? (() => Date.now());
+    const nowMs = nowMsFn();
+    const fiveMinutesAgoSec = Math.floor(nowMs / 1000) - 60 * 5;
+
+    // Enforce verification rules
+
+    // Rule 1: Check staleness
+    if (requestTimestampSec < fiveMinutesAgoSec) {
+      throw new Error(`${verifyErrorPrefix}: stale`);
+    }
+
+    // Rule 2: Check signature
+    // Separate parts of signature
+    const [signatureVersion, signatureHash] = signature.split('=');
+    // Only handle known versions
+    if (signatureVersion !== 'v0') {
+      throw new Error(`${verifyErrorPrefix}: unknown signature version`);
+    }
+    // Compute our own signature hash
+    const hmac = createHmac('sha256', signingSecret);
+    hmac.update(`${signatureVersion}:${requestTimestampSec}:${bufferedReq.rawBody.toString()}`);
+    const ourSignatureHash = hmac.digest('hex');
+    if (!tsscmp(signatureHash, ourSignatureHash)) {
```

Comment:
Does this throw a type error? `signatureHash` could be undefined.

Reply:
Probably? We can add more tests to check this